client/my-sites/email/email-providers-stacked-comparison/email-provider-stacked-card/email-provider-generic-form.tsx

```diff
@@ -0,0 +1,868 @@
+								onUserValueChange( 'password', event.target.value );
+							} }
+							onBlur={ () => {
+								setPasswordFieldTouched( wasValidated );
+							} }
+							onKeyUp={ onReturnKeyPress }
+						/>
+					</LabelWrapper>
+
+					{ hasPasswordError && <FormInputValidation text={ passwordError } isError /> }
+				</div>
+				{ showTrashButton && (
+					<Button
+						className="email-provider-generic-form__new-user-remove-user-button"
+						onClick={ onUserRemove }
+					>
+						<Gridicon icon="trash" />
+						<span>{ translate( 'Remove this mailbox' ) }</span>
+					</Button>
+				) }
+			</FormFieldset>
+		</div>
+	);
+};
+
+interface EmailProviderGenericFormProps {
+	autoFocus?: boolean;
+	children?: ReactNode;
+	domains?: any;
+	extraValidation: ( user: GenericNewUser ) => GenericNewUser;
+	selectedDomainName: string;
+	setValidForm: ( valid: boolean ) => void;
+	showAddAnotherMailboxButton: boolean;
+	onUsersChange: ( users: GenericNewUser[] ) => void;
+	onReturnKeyPress: ( event: Event ) => void;
+	hiddenFields: string[];
+	users: GenericNewUser[];
+}
+
+export const EmailProviderGenericForm: FunctionComponent< EmailProviderGenericFormProps > = ( {
+	autoFocus = false,
+	children,
+	domains,
+	extraValidation,
+	onUsersChange,
+	onReturnKeyPress,
+	hiddenFields = [],
+	selectedDomainName,
+	setValidForm,
+	showAddAnotherMailboxButton = true,
+	users,
+} ) => {
+	const translate = useTranslate();
+
+	const onUserValueChange = ( uuid: string ) => (
+		fieldName: string,
+		fieldValue: string,
+		mailBoxFieldTouched = false
+	) => {
+		const changedUsers = users.map( ( user ) => {
+			if ( user.uuid !== uuid ) {
+				return user;
+			}
+
+			const changedUser = { ...user, [ fieldName ]: { value: fieldValue, error: null } };
+
+			if ( GENERIC_EMAIL_FORM_FIRST_NAME_FIELD === fieldName && ! mailBoxFieldTouched ) {
+				return { ...changedUser, mailBox: { value: sanitizeEmail( fieldValue ), error: null } };
+			}
+
+			return changedUser;
+		} );
+
+		onUsersChange( validateUsers( changedUsers, extraValidation, hiddenFields ) );
+		setValidForm( areAllUsersValid( users, hiddenFields ) );
+	};
+
+	const onUserAdd = () => {
+		onUsersChange( [ ...users, newUser( selectedDomainName ) ] );
+	};
+
+	const onUserRemove = ( uuid: string ) => () => {
+		const newUserList = users.filter( ( _user ) => _user.uuid !== uuid );
+
+		onUsersChange( 0 < newUserList.length ? newUserList : [ newUser( selectedDomainName ) ] );
+	};
+
+	return (
+		<div>
+			{ users.map( ( user, index ) => (
+				<Fragment key={ user.uuid }>
+					<GenericNewUser
+						autoFocus={ autoFocus } // eslint-disable-line jsx-a11y/no-autofocus
+						domains={
+							domains ? domains.map( ( domain: any ) => domain.name ) : [ selectedDomainName ]
+						}
+						hiddenFields={ hiddenFields }
+						user={ user }
+						onUserValueChange={ onUserValueChange( user.uuid ) }
+						onUserRemove={ onUserRemove( user.uuid ) }
```

Comment:
Nit: these two functions return/build an event handler, so it may be clearer to call them something like `getUserValueChangeHandler()` and `getUserRemoveHandler()`.